Dashboard account notifications settings - add extras section

Addison-Stavlo opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of # https://linear.app/a8c/issue/CML-850/notifications-extras-screen

## Proposed Changes

* Adds a fetcher for `me/notifications/settings` to the `api-core` package, as well as a mutator for the wpcom (extras settings) data from this endpoint.
* Adds a query for me notifications settings to the `api-queries` package, as well as a mutation function for the extras settings.
* Updates the v2 dashboard's notification-extras component to contain cards with toggles for these WordPress.com and Jetpack settings.

BEFORE
<img width="1721" height="590" alt="Screenshot 2025-09-12 at 1 15 18 PM" src="https://github.com/user-attachments/assets/9868f830-8c16-4d6e-92a8-831085b38b98" />


AFTER
<img width="1748" height="1253" alt="Screenshot 2025-09-12 at 1 13 51 PM" src="https://github.com/user-attachments/assets/f99686d1-3c6b-4e13-b361-83ff17b6d60b" />



## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* Gotta fill in the new dashboards settings!

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to `*/v2/me/notifications/extras` for this build.
* Test the settings. Verify they update and persist as expected.
* Verify the top toggle for "Subscribe to all" / "Unsubscribe from all" work and appear in the correct cases (none selected => subscribe to all / any selected => unsubscribe)

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

Addison-Stavlo committed:
create me-notification-settings fetcher and mutator in api-core

Addison-Stavlo committed:
add notifications settings query and mutations to api-queries package

Addison-Stavlo committed:
add wpcom settings

Addison-Stavlo committed:
add jetpack toggles

Addison-Stavlo committed:
adjust spacing and types

Addison-Stavlo committed:
Merge branch 'trunk' into add/extras-scren-notifs-settings-v2-dash

Addison-Stavlo commented on client/dashboard/me/notifications-extras/index.tsx:
```diff
@@ -14,6 +183,92 @@ export default function NotificationsEmails() {
+					</CardBody>
+				</Card>
+
+				{ hasJetpackOptions && (
```
In the previous implementation, this rendering was conditional on the `hasJetpackSite` selector => only rendering this section if the user has a jetpack site.

I wonder if that logic is outdated and unnecessary here? as many Jetpack features are enabled on simple sites and the Jetpack section exists in the sidebar. I wonder if we should be showing this all the time? 🤔

gabrielcaires replied:
I don't have this answer as well. Can we poke someone about it?  Maybe the original code author.

Addison-Stavlo replied:
I am going to go with showing the jetpack section for everyone. I tested creating a brand new WPCom account and all the jetpack notifs are enabled by default without having a jetpack site at all:

<img width="358" height="390" alt="Screenshot 2025-09-15 at 12 48 23 PM" src="https://github.com/user-attachments/assets/5efcd8db-744b-46df-b995-f7a38ebee533" />

Meaning in our current state, and if we were to continue to hide based on presence of having a jetpack site, users would not actually be able to turn these off in the settings without actually creating a Jetpack site first. So this seems like a bit of a bug that evolved that we can correct here.

Addison-Stavlo committed:
remove savableKeys that we are not managing with these settings

gabrielcaires commented on client/dashboard/me/notifications-extras/index.tsx:
```diff
@@ -1,8 +1,177 @@
+};
+
+export default function NotificationsExtras() {
+	const { data } = useQuery( {
```
We prefer to use `useSuspenseQuery` because it prevents layout shifts while the request is still happening.

gabrielcaires commented on client/dashboard/me/notifications-extras/index.tsx:
```diff
@@ -14,6 +183,92 @@ export default function NotificationsEmails() {
+						<SectionHeader level={ 3 } title={ __( 'Email from WordPress.com' ) } />
+
+						<Spacer marginTop={ 8 }>
+							<ToggleControl
```
We want to use DataForm from the `@wordpress/dataview` as much as we can; it will probably reduce a lot of the code.  You can see an example [here](https://github.com/Automattic/wp-calypso/blob/eab1aa11755dc9fd52b16012092ae463a92539a7/apps/design-system-docs/docs/components/packages/wp-components/wp-components-table.tsx#L9).

Addison-Stavlo replied:
Im not sure I follow the linked example? These are for basic individual toggles and Im not sure how a DataView to sort, filter, paginate, and set view type makes sense here?

fushar replied:
Since the toggles are auto-saving, I'm not sure if we can use DataForm here... I'll let you folks figure out 🙈

gabrielcaires replied:
Ops, wrong link...
I was mentioning dataform https://wordpress.github.io/gutenberg/?path=/docs/dataviews-dataform--docs

gabrielcaires replied:
Thinking a bit better, I am facing relevant issues with DataForm, like this https://github.com/WordPress/gutenberg/issues/71616, this https://github.com/Automattic/wp-calypso/pull/105693#issuecomment-3286447292, and another that I will still publish on the gutenberg repo.

If you agree, for now, we can focus on addressing the other feedback and create an additional project task to evaluate whether it makes sense to convert this and https://github.com/Automattic/wp-calypso/pull/105571 to use DataForm.

gabrielcaires commented on packages/api-core/src/me-notification-settings/mutators.ts:
```diff
@@ -0,0 +1,29 @@
+export async function updateWpcomNotificationSettings(
+	data: Partial< WpcomNotificationSettings >
+): Promise< Partial< WpcomNotificationSettings > > {
+	const saveableKeys: ( keyof WpcomNotificationSettings )[] = [
```
We can remove it.

p1757427633721609/1757427287.867389-slack-C0982082MSR removing these keys and leaving each UI implementation to define which properties should be updated.

gabrielcaires commented on client/dashboard/me/notifications-extras/index.tsx:
```diff
@@ -14,6 +183,92 @@ export default function NotificationsEmails() {
+					<CardBody>
+						<SectionHeader level={ 3 } title={ __( 'Email from WordPress.com' ) } />
+
+						<Spacer marginTop={ 8 }>
```
It would be nice if we could avoid using the spacer component. If we split the controls into two groups (subscribe to all, others), we can use a vspace with the correct gap. 

The manual margin top is hard to maintain in future design system evolutions.

gabrielcaires reviewed (changes requested):
I left some comments about the implementation details. I also suggest we split the big file into smaller modules and add at least a few automated tests. It can ensure that we and other teams can continue evolving faster.

fushar commented on packages/api-queries/src/me-notifications-settings.ts:
```diff
@@ -0,0 +1,28 @@
+import { queryOptions, mutationOptions } from '@tanstack/react-query';
+import { queryClient } from './query-client';
+
+export const meNotificationsSettingsQuery = () =>
```
[nit] for consistency with other `me-*` queries, let's use `userNotificationSettingsQuery()` rather than `meNotificationSettingsQuery()` (also the mutation)

fushar commented on packages/api-core/src/me-notification-settings/types.ts:
```diff
@@ -0,0 +1,31 @@
+	gravatar_onboarding: boolean;
+}
+
+export interface MeNotificationSettings {
```
Same (`UserNotificationSettings`)

gabrielcaires commented on client/dashboard/me/notifications-extras/index.tsx:
```diff
@@ -1,8 +1,177 @@
+		...meNotificationsSettingsQuery(),
+		meta: { persist: false },
+	} );
+	const mutation = useMutation( meNotificationsExtrasSettingsMutation() );
```
We can add a notification to confirm that the save was successful.
I recently learned how to do it here.
 
(https://github.com/Automattic/wp-calypso/pull/105693/files#diff-187240df604a75e07301065bd0042bf47aba8e71e232062d938942e2b16c63d9R42-R44)

Addison-Stavlo committed:
use the suspense query

Addison-Stavlo committed:
remove allow list from mutator

Addison-Stavlo committed:
remove use of spacers, rely on VStack

Addison-Stavlo committed:
update naming conventions

Addison-Stavlo committed:
refactor a shared component to handle toggles and logic

Addison-Stavlo committed:
remove unnecessary key check for jetpack

Addison-Stavlo committed:
Merge branch 'trunk' into add/extras-scren-notifs-settings-v2-dash

Addison-Stavlo committed:
Merge branch 'trunk' into add/extras-scren-notifs-settings-v2-dash

Addison-Stavlo commented:
Ok, this should be ready for another look. I consolidated a lot of the logic out of the large file to reuse between both cards, and addressed all of the above feedback except the following 2 items:

> We can add a notification to confirm that the save was successful.
I recently learned how to do it here.

That sounds great! But I think it will be more simple to follow up and add that once you have merged the noted PRs and handled the noted unification you mention around the useNotice hook. It should be a simple addition once thats ready, I can create a follow-up issue to track this.

> a few automated tests

I haven't gotten to this yet. If this approach looks like a reasonable way to go then Il be happy to look at adding some. Note I may not get a chance to do much more regarding that for a couple days -- So I could add this in another follow up, or we stall this PR for a couple days until we get a chance to add those. I may prefer the former but really either way works.

fushar reviewed (commented):
Code-wise I think it's good; did not test the feature; I'll let Loop team to do it 😄

gabrielcaires reviewed (approved):
Worked as expected. I think it's a bit unusual to have all the buttons on while we still have settings off, but I saw that the current implementation is doing it


We can add the tests in a another PR.

gabrielcaires committed:
Merge branch 'trunk' into add/extras-scren-notifs-settings-v2-dash

gabrielcaires commented:
@Addison-Stavlo, I merged the PR because I need part of the code.
